File: R/run_l63_var.R
source("l63.R")
source("step.R")
source("config_l63.R")

seed <- 514
set.seed(seed)

calc.cost <- function(dx, b, dy, r) {
  cost <- 0.5 * t(dx) %*% dx / b
  for (j in 1:ncol(dy)) {
    cost <- cost + 0.5 * t(dy[, j]) %*% (dy[, j] / r)
  }
  as.numeric(cost)
}
options(digits = 5)
xt <- matrix(rep(0, ns * nt), nrow=ns)
xt[, 1] <- xt0
yo <- matrix(rep(0, ns * nobs), nrow=ns)
m <- 1
for (k in 2:nt) {
  xt[, k] <- step.fom(l63, xt[, k-1], 1, dt, p, r, b)
  if (k %% obs.int == 0) {
    yo[, m] <- rnorm(ns, xt[, k], sqrt(obs.r))
    m <- m + 1
  }
}

l2 <- rep(0, ni)
cost <- rep(0, ni)
xb <- matrix(rep(0, ns * nt), nrow=ns)
xb[, 1] <- xb0

for (i in 1:ni) {
  for (j in 2:nt) {
    xb[, j] <- step.fom(l63, xb[, j-1], 1, dt, p, r, b)    
  }
  d <- xb[, (1:nobs) * obs.int] - yo
  ad <- rep(0, ns)
  m = nobs
  for (j in nt:1){
    ad <- step.adm(l63, l63.ad, xb[, j], ad, 1, dt, p, r, b)
    if (j %% obs.int == 0) {
      ad <- ad + d[, m] / obs.r
      m <- m - 1
    }
  }
  xb[, 1] <- xb[, 1] - a * ad
  l2[i] <- sqrt(mean((xb[, 1] - xt0)^2))
  cost[i] <- calc.cost(xb[,1] - xb0, bgd.b, d, obs.r)
  cat(i, l2[i], cost[i], "\n")
}

tab10.new <- c('#5778a4', '#e49444', '#d1615d', '#85b6b2', '#6a9f58',
               '#e7ca60', '#a87c9f', '#f1a2a9', '#967662', '#b8b0ac')

title <- paste("L63 Var a=", a, " ni=", ni, " L2=", format(l2[ni], digits=3))
off <- c(0, 40, 60)
t <- (1:nt) * dt
to <- ((1:nobs) * obs.int) * dt
xt <- xt + off
xb <- xb + off
yo <- yo + off
plot(1, type = "n", xlab="time", ylab="state", main=title,
     xlim=c(min(t), max(t)), ylim=c(-20, 150))
for (i in 1:ns) {
  lines(t, xt[i,], lwd=2, lty=2, col=tab10.new[i])
  lines(t, xb[i,], lwd=2, col=tab10.new[i])
  points(to, yo[i,], cex=2, pch=4, col=tab10.new[i])
  abline(h=off[i], lty=3, col=tab10.new[i])
}
legend("topright", legend=c("model", "true", "obs"), 
       lty=c(1, 2, NA), pch=c(NA, NA, 4), ncol=3)

File: Python/run_l63_var.py
import numpy as np
import matplotlib.pyplot as plt
import l63
import step
from config_l63 import xt0, xb0, ns, p, r, b, dt, \
        nt, model_q, obs_int, obs_r, nobs, bgd_b, ni, a 
import sys


seed = 514
rng = np.random.default_rng(seed)

def calc_cost(dx, b, dy, r):
    cost = 0.5 * np.dot(dx, dx) / b
    for j in range(dy.shape[1]):
        cost += 0.5 * np.dot(dy[:, j], dy[:, j] / r)
    return cost

xt = np.zeros([ns, nt])
xt[:, 0] = xt0
yo = np.zeros([ns, nobs])
m = 0
for k in range(1,nt):
    xt[:, k] = step.fom(l63.l63, xt[:, k-1], 1, dt, p, r, b)
    if (k+1) % obs_int == 0:
        yo[:, m] = rng.normal(xt[:, k], np.sqrt(obs_r), ns)
        m += 1

l2 = np.zeros(ni)
cost = np.zeros(ni)
xb = np.zeros([ns, nt])
xb[:, 0] = xb0
for i in range(ni):
    m = 0
    for j in range(1, nt):
        xb[:, j] = step.fom(l63.l63, xb[:, j-1], 1, dt, p, r, b)    
    d = xb[:, (np.arange(nobs) + 1) * obs_int - 1] - yo
    ad = np.zeros(ns)
    m = nobs - 1
    for j in range(nt-1, -1, -1):
        ad = step.adm(l63.l63, l63.ad, xb[:, j], ad, 1, dt, p, r, b)
        if (j+1) % obs_int == 0:
            ad += d[:, m] / obs_r
            m -= 1
    xb[:, 0] = xb[:, 0] - a * ad
    l2[i] = np.sqrt(((xb[:, 0] - xt0)**2).mean())
    cost[i]  = calc_cost(xb[:, 0] - xb0, bgd_b, d, obs_r)
    print(f"i={i} L2={l2[i]:.3e} J={cost[i]:.3e}")

tab = ["tab:blue", "tab:orange", "tab:red"]

title = f"L63 Var a={a:.2e} ni={ni} L2={l2[-1]:.3e}"
off = np.array([0, 40, 60])

t = np.arange(1, nt+1) * dt
to = (np.arange(1,nobs+1) * obs_int) * dt
xt = xt + off[:, None]
xb = xb + off[:, None]
yo = yo + off[:, None]
fig, ax = plt.subplots()
for i in range(ns):
    l_xt = ax.plot(t, xt[i,], lw=2, ls="--", c=tab[i])
    l_xb = ax.plot(t, xb[i,], lw=2, c=tab[i])
    l_yo = ax.scatter(to, yo[i,], s=100, marker="x", c=tab[i])
    ax.axhline(off[i], ls=":", c=tab[i])
ax.set_title(title)
ax.set_xlabel("time")
ax.set_ylabel("state")
ax.set_xlim([t[0], t[-1]])
ax.set_ylim([-20, 150])
plt.show()
